sw/device/lib/dif/dif_i2c.h

```diff
@@ -0,0 +1,384 @@
+ * once timings, interrupts, and watermarks are all configured.
+ *
+ * @param i2c An I2C device.
+ * @param state The enablement state for the host functionality.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_host_set_enabled(const dif_i2c_t *i2c,
+                                          dif_i2c_enable_t state);
+
+/**
+ * Enables or disables the "override mode". In override mode, software is able
+ * to directly control the driven values of the SCL and SDA lines using
+ * `dif_i2c_override_drive_pins()`.
+ *
+ * @param i2c An I2C device.
+ * @param state The enablement state for override mode.'
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_override_set_enabled(const dif_i2c_t *i2c,
+                                              dif_i2c_enable_t state);
+
+/**
+ * Drives the SCL and SDA pins to the given values when "override mode" is
+ * enabled.
+ *
+ * @param i2c An I2C device.
+ * @param scl The value to drive SCL to.
+ * @param sda The value to drive SDA to.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_override_drive_pins(const dif_i2c_t *i2c, bool scl,
+                                             bool sda);
+
+/**
+ * Returns oversampling of the last 16 values of the SCL and SDA pins, with the
+ * zeroth bit being the most recent.
+ *
+ * @param i2c An I2C device.
+ * @param scl_samples Out-param for SCL sample bits; may be null.
+ * @param sda_samples Out-param for SDA sample bits; may be null.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_override_sample_pins(const dif_i2c_t *i2c,
+                                              uint16_t *scl_samples,
+                                              uint16_t *sda_samples);
+
+/**
+ * Returns the current levels, i.e., number of entries, in the FMT and RX FIFOs.
+ * These values represent the number of entries pending for send by hardware,
+ * and entries pending for read by software, respectively.
+ *
+ * @param i2c An I2C device.
+ * @param fmt_fifo_level Out-param for the number of unsent FMT bytes; may be
+ *        null.
+ * @param rx_fifo_level Out-param for the number of unread RX bytes; may be
+ *        null.
+ * @return The result of the operation.
+ */
+dif_i2c_result_t dif_i2c_get_fifo_levels(const dif_i2c_t *i2c,
+                                         uint8_t *fmt_fifo_level,
+                                         uint8_t *rx_fifo_level);
+
+/**
+ * Pops an entry (a byte) off of the RX FIFO. Passing in `NULL` to the out-param
+ * will still trigger a byte pop.
+ *
+ * @param i2c An I2C device.
+ * @param byte Out-param for the popped byte; may be null.
+ * @return The result of the opeartion.
+ */
+dif_i2c_result_t dif_i2c_read_byte(const dif_i2c_t *i2c, uint8_t *byte);
+
+/**
+ * Flags for a formatted I2C byte, used by the `dif_i2c_write_byte_raw()`
+ * function.
+ */
+typedef struct dif_i2c_fmt_flags {
```

Comment:
That feels vaguely like a driver, rather than DIF, API? If we want to keep the `raw` variant, which does flag checking anyway, I don't think there's any reason to add something mildly more complex on top.

Also, reading this code again, I wonder if the struct actually wants to be a flag enum. This connects back to something I said a while back about interrupt type enums. @lenary, what do you think?

Reply:
I agree with Alphan. If you look at the "Implementation Details: Format Flag Parsing"
section of the IP docs then there's a clear set of valid transitions. The enum that
@alphan is suggesting ensures that the bundle of flags are valid are exposed.

Reply:
It seems like you agreed but didn't change the interface. :wink:

Reply:
+1